apps/game_backend/lib/game_backend/curse_of_mirra/quests.ex

```diff
@@ -298,6 +296,79 @@ defmodule GameBackend.CurseOfMirra.Quests do
     |> Repo.transaction()
   end
 
+  def get_user_quest_progress(%UserQuest{quest: %Quest{} = quest} = user_quest, arena_match_results) do
+    arena_match_results =
+      case quest.type do
+        type when type in ["daily", :daily] ->
+          Enum.filter(arena_match_results, fn arena_match_result ->
+            user_quest.activated_at &&
+              NaiveDateTime.compare(arena_match_result.inserted_at, user_quest.activated_at) == :gt &&
+              NaiveDateTime.diff(arena_match_result.inserted_at, user_quest.inserted_at, :day) == 0
+          end)
+
+        _ ->
+          arena_match_results
+      end
+
+    arena_match_results
+    |> filter_results_that_meet_quest_conditions(quest.conditions)
+    |> Enum.reduce(0, fn arena_match_result, acc ->
+      type = arena_match_result_field_to_atom(quest.objective["match_tracking_field"])
+      # We won't accumulate progress if the objective has a wrong field getter
+      accumulate_objective_progress_by_scope(quest.objective["scope"], Map.get(arena_match_result, type))
+      |> case do
+        nil ->
+          acc
+
+        progress ->
+          acc + progress
+      end
+    end)
+  end
+
+  def try_to_complete_quest_for_user(user, user_quest) do
+    updated_user_quest_changeset =
+      UserQuest.changeset(user_quest, %{
+        completed: true,
+        completed_at: DateTime.utc_now(),
+        status: "completed"
+      })
+
+    Multi.new()
+    |> Multi.run(:check_quest_completed, fn _, _ ->
+      if user_quest.status == "available" && Quests.completed_quest?(user_quest, user.arena_match_results) do
+        {:ok, :quest_completed}
+      else
+        {:error, :unfinished_quest}
+      end
+    end)
+    |> Multi.update(:update_user_quest, updated_user_quest_changeset)
+    |> Multi.run(:maybe_activate_quest, fn _, _ ->
+      Enum.find(user.user_quests, fn user_quest -> is_nil(user_quest.activated_at) end)
+      |> case do
+        nil ->
+          {:ok, :no_quests_left}
+
+        user_quest ->
+          user_quest
+          |> UserQuest.changeset(%{activated_at: NaiveDateTime.utc_now()})
+          |> GameBackend.Repo.update()
```

Comment:
Do we want to do this everytime? Shouldn't we isolate this behavior for weekly ones only?
Are bounty and daily ones under this "activate" behavior too?
What happens if I complete a bounty. Will that activate a weekly?

Reply:
1. This is for the daily quests
2. Since we don't have a specification on how the weekly quests will behave I'll just add a type filter so we don't mess up the different quests